config_converter/config_mapper/config_mapper.py

```diff
@@ -0,0 +1,163 @@
+    # these dicts can be updated when more plugins are supported
+    for d in config_obj.directives:
+        if d.name not in plugin_prefix_map:
+            print(f'Currently we do not support plugins of {d.name}')
+            continue
+        try:
+            plugin_type = next(p.value for p in d.params if p.name == '@type')
+        except StopIteration:
+            print('Invalid configuration - missing @type param')
+            sys.exit()
+        plugin_name = plugin_prefix_map[d.name] + plugin_type
+        if plugin_name not in _SUPPORTED_PLUGINS:
+            print(f'We do not know plugin {plugin_name}')
+        else:
+            plugin_dir = dir_name_map[d.name]
+            if plugin_dir not in logs_module:
+                logs_module[plugin_dir] = []
+            logs_module[plugin_dir].append(_convert_plugin(d, plugin_name))
+    return {'logs_module': logs_module}
+
+
+def _convert_plugin(d: config_pb2.Directive, plugin: str) -> dict:
+    """Cases on plugin, calls corresponding mapping function."""
+    result = dict()
+    plugin_type_map = {'tail': 'file'}
+    plugin_type = next(p.value for p in d.params if p.name == '@type')
+    result['type'] = plugin_type_map[plugin_type]
+    try:
+        result['name'] = next(p.value for p in d.params if p.name == 'tag')
+    except StopIteration:
+        print('Invalid configuration - missing tag')
+        sys.exit()
+    if plugin == 'in_tail':
+        result[f'{result["type"]}_{d.name}_config']: dict = _convert_in_tail(d)
+    return result
+
+
+def _convert_in_tail(d: config_pb2.Directive) -> dict:
+    """Uses mapping rules to convert in_tail plugin fields."""
+    fields = dict()
+    # these fields may not belong to the same level, have a 1:1 mapping, etc
+    # https://docs.fluentd.org/parser/multiline - shows formatN works for
+    # 1 <= N <= 20
+    special_fields = [
+        'format', 'format_firstline', '@type', 'multiline_flush_interval',
+        'expression'
+    ] + [f'format{i}' for i in range(1, 21)]
+    for p in d.params:
+        if p.name == '@type' or p.name == 'tag':
+            continue
+        if p.name == 'exclude_path':
+            fields['exclude_path'] = p.value
+        elif p.name == 'path':
+            fields['path'] = p.value
+        elif p.name == 'path_key':
+            fields['path_field_name'] = p.value
+        elif p.name == 'pos_file':
+            fields['checkpoint_file'] = p.value
+        elif p.name == 'refresh_interval':
+            fields['refresh_interval'] = p.value
+        elif p.name == 'rotate_wait':
+            fields['rotate_wait'] = p.value
+        elif p.name in special_fields:
+            _convert_parse_dir(fields, p)
+        elif p.name in _UNSUPPORTED_FIELDS:
+            print(f'{p.name} cannot be mapped into master agent config file.')
+        else:
+            print(f'{p.name} is an unknown field.')
+    for nd in d.directives:
+        if nd.name == 'parse':
+            for np in nd.params:
```

Comment:
Could we structure the parser converter similarly to the in_tail converter? I'm thinking of code like this:

    if nd.name == 'parse':
      fields['parser'] = _convert_parse_dir(nd)

Would that work?

Reply:
I initially wanted to do that too - it's just that then I'd have to write more code for parameters like format_firstline, format, etc if not put within a parser directive. The logic is identical to the parser function, so it would basically be lots of repetitive code. Is there another way to make this simpler maybe?